gctree/branching_processes.py

```diff
@@ -987,3 +964,230 @@ def f(x):
     if not result.success:
         warnings.warn("optimization not sucessful, " + result.message, RuntimeWarning)
     return result.x[0], result.x[1]
+
+
+def lltree(cm_counts, p: np.float64, q: np.float64) -> Tuple[np.float64, np.ndarray]:
+    r"""Log likelihood of branching process parameters :math:`(p, q)`
+    .. math::
+        \ell(p, q; T, A) = \log\mathbb{P}(T, A \mid p, q)
+
+    Args:
+        cm_counts: an iterable containing tuples `((c, m), n)` where `n` is the number of nodes
+            in the tree with abundance `c` and `m` mutant clades
+        p: branching probability
+        q: mutation probability
+    Returns:
+        Log likelihood :math:`\ell(p, q; T, A)` and its gradient :math:`\nabla\ell(p, q; T, A)`
+    """
+    count_ls = [n for cm, n in cm_counts]
+    cm_list = [cm for cm, n in cm_counts]
+    if cm_list[0][0] == 0 and cm_list[0][1] == 1:
+        # if unifurcation not possible under current model, add a
+        # psuedocount for the root
+        cm_list[0] = (1, 1)
+    # extract vector of function values and gradient components
+    logf_data = [CollapsedTree._ll_genotype(c, m, p, q) for c, m in cm_list]
+    logf = np.array([[x[0] * count_ls[i]] for i, x in enumerate(logf_data)]).sum()
+    grad_ll_genotype = np.array(
+        [x[1] * count_ls[i] for i, x in enumerate(logf_data)]
+    ).sum(axis=0)
+    return logf, grad_ll_genotype
+
+
+def llforest(
```

Comment:
Does this need to be a public function in the API that is shown in docs? If not, suggest renaming `_llforest`.